### PR TITLE
feat: cosmos DeFi opportunity modals multi-account support

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

- [x] Overview account selection
- [x] Deposit account selection across all steps
- [x] Withdraw account selection
- [x] Broadcast deposits/withdraws Txs from the selected account
- [x] Available amounts staked in overview / available to deposit/withdraw are matching the ones for the account
- [x] Claim amount for the selected account
- [x] Regression test with the flag off
- [ ] FOUC on deposit/withdraw modals - investigate reconciliation of `<AccountDropdown />`
- [ ] ~~Confirm Osmosis is working~~ - not for this PR since Osmosis multi account isn't properly working currently
- [ ] ~~Deep-linking - use query state instead of props drilling?~~ TBD in https://github.com/shapeshift/web/issues/2615

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/blob/develop/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/2606

## Risk

Full regression test needed with the flag off, logic has been revamped to account for multi-account.

#### With the `MultiAccounts` flag on

- Add a Cosmos account and send some ATOM to it
- Open the Cosmos overview modal for a validator you're currently delegating to on account #0, select account #1 and click "Deposit"
- Notice account #1 is pre-selected at Deposit step, with the right amount to deposit 
- Deposit some ATOM - ATOM should be deposited from account #1
- Disconnect and reconnect wallet, then reopen the Cosmos overview modal and select withdraw with account #1 selected
- Notice account #1 is pre-selected at Withdraw step, with the right amount available to withdraw
- Withdraw some ATOM - ATOM should be withdrawn from account #1
- Disconnect and reconnect wallet, then reopen the Cosmos overview modal with account #1 selected and notice the undelegation is shown in the unstaking list
- Click "Claim" and notice the claim is TBD from account #1, with the right amount to claim (most likely 0 because of the precision we use, if you try a few blocks later you should see some actual displayed amount - confirmed through console debugging the amount to claim is correct)
- Claim should go through from account #1

### Engineering

- Refer to top-level testing steps

### Operations

- Refer to top-level testing steps
- Regression test with the `MultiAccounts` flag off, making sure to disconnect and reconnect your native wallet when turning the flag on and off.

#### No account added

image
image
image

#### Second Cosmos account added

image
image
image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #2728** <a href="https://app.graphite.dev/github/pr/shapeshift/web/2728" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/2728?utm_source=stack-comment).

File: src/components/AccountDropdown/AccountDropdown.tsx [hunk 1]
```diff
@@ -110,6 +112,7 @@ export const AccountDropdown: FC<AccountDropdownProps> = ({
    * react on accountIds on first render
    */
   useEffect(() => {
+    debugger
```
Comment (gomesalexandre) [inline src/components/AccountDropdown/AccountDropdown.tsx:115]: Remove before opening PR

File: src/components/AccountDropdown/AccountDropdown.tsx [hunk 2]
```diff
@@ -247,7 +250,10 @@ export const AccountDropdown: FC<AccountDropdownProps> = ({
    * the effectful logic above will still run for other chains, and return the first account
    * via the onChange callback on mount, but nothing will be visually rendered
    */
-  const existingMultiAccountChainIds = useMemo(() => [btcChainId, ltcChainId], [])
+  const existingMultiAccountChainIds = useMemo(
+    () => [btcChainId, ltcChainId, cosmosChainId, osmosisChainId],
+    [],
+  )
```
Comment (gomesalexandre) [inline src/components/AccountDropdown/AccountDropdown.tsx:256]: Revert

File: src/features/defi/providers/cosmos/components/CosmosManager/Claim/CosmosClaim.tsx [hunk 3]
```diff
@@ -31,7 +33,7 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'CosmosClaim'],
 })
 
-export const CosmosClaim = () => {
+export const CosmosClaim: React.FC<{ accountId?: Nullable<AccountId> }> = ({ accountId }) => {
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Claim/CosmosClaim.tsx:36]: No-inline

File: src/features/defi/providers/cosmos/components/CosmosManager/Deposit/CosmosDeposit.tsx [hunk 4]
```diff
@@ -75,7 +85,7 @@ export const CosmosDeposit = () => {
         const chainAdapter = chainAdapterManager.get(chainId)
         if (!(walletState.wallet && contractAddress && chainAdapter && apr)) return
```
Comment (0xdef1cafe) [inline src/features/defi/providers/cosmos/components/CosmosManager/Deposit/CosmosDeposit.tsx:86]: && bip44params

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/CosmosWithdraw.tsx [hunk 5]
```diff
@@ -76,11 +85,15 @@ export const CosmosWithdraw = () => {
...
   ) as unknown as MergedActiveStakingOpportunity // TODO: remove casting
 
+  const accountFilter = useMemo(() => ({ accountId: accountId ?? '' }), [accountId])
+  const bip44Params = useAppSelector(state => selectBIP44ParamsByAccountId(state, accountFilter))
+
   useEffect(() => {
+    if (!bip44Params) return
     ;(async () => {
       try {
         if (!(walletState.wallet && contractAddress && chainAdapter)) return
```
Comment (0xdef1cafe) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/CosmosWithdraw.tsx:95]: && bip44Params

File: src/plugins/cosmos/hooks/useStakingAction/useStakingAction.tsx [hunk 6]
```diff
@@ -42,7 +43,7 @@ export const useStakingAction = () => {
   } = useWallet()
 
   const handleStakingAction = async (data: StakingInput) => {
-    const { chainSpecific, validator, action, value, asset } = data
+    const { bip44Params, chainSpecific, validator, action, value, asset } = data
 
     if (!wallet) return
```
Comment (0xdef1cafe) [inline src/plugins/cosmos/hooks/useStakingAction/useStakingAction.tsx:48]: && bip44Params

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Withdraw.tsx [hunk 7]
```diff
@@ -33,7 +35,12 @@ export type CosmosWithdrawValues = {
 
 const moduleLogger = logger.child({ namespace: ['CosmosWithdraw:Withdraw'] })
 
-export const Withdraw: React.FC<StepComponentProps> = ({ onNext }) => {
+export const Withdraw: React.FC<
```
Comment (0xdef1cafe) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Withdraw.tsx:38]: no inline

File: src/features/defi/providers/cosmos/components/CosmosManager/Claim/components/Confirm.tsx [hunk 8]
```diff
@@ -31,7 +37,10 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'Claim', 'Confirm'],
 })
 
-export const Confirm = ({ onNext }: StepComponentProps) => {
+export const Confirm = ({
+  accountId,
+  onNext,
+}: StepComponentProps & { accountId?: Nullable<AccountId> }) => {
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Claim/components/Confirm.tsx:43]: No-inline, `ConfirmProps`

File: src/features/defi/providers/cosmos/components/CosmosManager/Deposit/CosmosDeposit.tsx [hunk 9]
```diff
@@ -39,7 +42,10 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'CosmosDeposit'],
 })
 
-export const CosmosDeposit = () => {
+export const CosmosDeposit: React.FC<{
+  onAccountIdChange: AccountDropdownProps['onChange']
+  accountId: AccountId | null
+}> = ({ onAccountIdChange: handleAccountIdChange, accountId }) => {
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Deposit/CosmosDeposit.tsx:48]: No-inline, `CosmosDepositProps`

File: src/features/defi/providers/cosmos/components/CosmosManager/Deposit/components/Confirm.tsx [hunk 10]
```diff
@@ -35,7 +37,10 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'Deposit', 'Confirm'],
 })
 
-export const Confirm = ({ onNext }: StepComponentProps) => {
+export const Confirm: React.FC<StepComponentProps & { accountId?: AccountId | null }> = ({
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Deposit/components/Confirm.tsx:40]: No-inline, `ConfirmProps`

File: src/features/defi/providers/cosmos/components/CosmosManager/Deposit/components/Deposit.tsx [hunk 11]
```diff
@@ -11,23 +12,30 @@ import { getFormFees } from 'plugins/cosmos/utils'
...
 
 const moduleLogger = logger.child({ namespace: ['CosmosDeposit:Deposit'] })
 
-export const Deposit: React.FC<StepComponentProps> = ({ onNext }) => {
+export const Deposit: React.FC<
+  StepComponentProps & {
+    accountId: Nullable<AccountId>
+    onAccountIdChange: AccountDropdownProps['onChange']
+  }
+> = ({ onNext, accountId, onAccountIdChange: handleAccountIdChange }) => {
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Deposit/components/Deposit.tsx:38]: No-inline, `DepositProps`

File: src/features/defi/providers/cosmos/components/CosmosManager/Overview/CosmosOverview.tsx [hunk 12]
```diff
@@ -29,11 +31,15 @@ import {
 } from 'state/slices/selectors'
 import { getDefaultValidatorAddressFromAssetId } from 'state/slices/validatorDataSlice/utils'
 import { useAppSelector } from 'state/store'
+import type { Nullable } from 'types/common'
 
 import { CosmosEmpty } from './CosmosEmpty'
 import { WithdrawCard } from './WithdrawCard'
 
-export const CosmosOverview = () => {
+export const CosmosOverview: React.FC<{
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Overview/CosmosOverview.tsx:39]: No-inline, `CosmosOverviewProps`

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/CosmosWithdraw.tsx [hunk 13]
```diff
@@ -33,7 +39,10 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'CosmosWithdraw'],
 })
 
-export const CosmosWithdraw = () => {
+export const CosmosWithdraw: React.FC<{
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/CosmosWithdraw.tsx:42]: No-inline, `React.FC<CosmosWithdrawProps>`

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Confirm.tsx [hunk 14]
```diff
@@ -38,7 +40,10 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'Withdraw', 'Confirm'],
 })
 
-export const Confirm = ({ onNext }: StepComponentProps) => {
+export const Confirm: React.FC<StepComponentProps & { accountId: AccountId | null }> = ({
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Confirm.tsx:43]: No-inline, `React.FC<ConfirmProps>`

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Withdraw.tsx (hunk 7 above)
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Withdraw.tsx:38]: No-inline, `WithdrawProps`

**Review by gomesalexandre — COMMENTED** [review]
Self-review, use `React.FC` and abstract props into their own type vs. inlining them

File: src/features/defi/providers/cosmos/components/CosmosManager/Overview/CosmosOverview.tsx [hunk 15]
```diff
@@ -29,14 +31,28 @@ import {
...
+}) => {
   const translate = useTranslate()
   const { query, history, location } = useBrowserRouter<DefiQueryParams, DefiParams>()
-  const { chainId, contractAddress, assetReference } = query
+  const { defaultAccountId: queryAccountId, chainId, contractAddress, assetReference } = query
+
+  const accountId = useMemo(
+    () => defaultAccountId ?? queryAccountId,
+    [defaultAccountId, queryAccountId],
+  )
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Overview/CosmosOverview.tsx:55]: This is the guts of d02319b5fe21617053f49ecf3932a12c8a786838 - derp-linking paving the way for deep-linking and removal of `accountSpecifier`.

We will be able to either be fully controlled at manager-level, or get the accountId from queryparams, then be controlled at manager-level

File: src/features/defi/providers/cosmos/components/CosmosManager/CosmosManager.tsx [hunk 16]
```diff
@@ -16,6 +18,8 @@ import { CosmosWithdraw } from './Withdraw/CosmosWithdraw'
 export const CosmosManager = () => {
   const { query, history: browserHistory } = useBrowserRouter<DefiQueryParams, DefiParams>()
   const { modal } = query
+  const [accountId, setAccountId] = useState<AccountId | null>(null)
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/CosmosManager.tsx:21]: Nullable

File: src/features/defi/providers/cosmos/components/CosmosManager/Deposit/CosmosDeposit.tsx [hunk 17]
```diff
@@ -39,7 +42,15 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'CosmosDeposit'],
 })
 
-export const CosmosDeposit = () => {
+type CosmosDepositProps = {
+  onAccountIdChange: AccountDropdownProps['onChange']
+  accountId: AccountId | null
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Deposit/CosmosDeposit.tsx:47]: Nullable

File: src/features/defi/providers/cosmos/components/CosmosManager/Deposit/components/Confirm.tsx [hunk 18]
```diff
@@ -35,7 +37,9 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'Deposit', 'Confirm'],
 })
 
-export const Confirm = ({ onNext }: StepComponentProps) => {
+type ConfirmProps = StepComponentProps & { accountId?: AccountId | null }
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Deposit/components/Confirm.tsx:40]: Nullable

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/CosmosWithdraw.tsx [hunk 19]
```diff
@@ -33,7 +39,14 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'CosmosWithdraw'],
 })
 
-export const CosmosWithdraw = () => {
+type CosmosWithdrawProps = {
+  accountId: AccountId | null
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/CosmosWithdraw.tsx:43]: Nullable

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Confirm.tsx [hunk 20]
```diff
@@ -38,7 +40,9 @@ const moduleLogger = logger.child({
   namespace: ['DeFi', 'Providers', 'Cosmos', 'Withdraw', 'Confirm'],
 })
 
-export const Confirm = ({ onNext }: StepComponentProps) => {
+type ConfirmProps = StepComponentProps & { accountId: AccountId | null }
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Confirm.tsx:43]: Nullable

File: src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Withdraw.tsx [hunk 21]
```diff
@@ -33,7 +35,15 @@ export type CosmosWithdrawValues = {
 
 const moduleLogger = logger.child({ namespace: ['CosmosWithdraw:Withdraw'] })
 
-export const Withdraw: React.FC<StepComponentProps> = ({ onNext }) => {
+type WithdrawProps = StepComponentProps & {
+  accountId: AccountId | null
```
Comment (gomesalexandre) [inline src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/components/Withdraw.tsx:39]: Nullable

**Review by 0xdef1cafe — DISMISSED** [review]

**Comment by 0xdef1cafe** [comment]
@gomesalexandre stamped - feel free to fix conflict and admin merge

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**